runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,325 @@
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.count();
+    }
+
+    /**
+     * Gets all the keys of a table.
+     *
+     * @param table - the table whose keys are requested.
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the corfu transaction. When not reading transactional writes, the internal
+     * corfu transaction will begin on transaction commit()
+     */
+    private boolean txBeginMaybe() {
+        if (TransactionalContext.isInTransaction()) {
+            return true;
+        }
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+        return false;
+    }
+
+    private long txEndInternal() {
+        if (TransactionalContext.isInTransaction()) {
+            return this.objectsView.TXEnd();
+        }
+        return Address.NON_ADDRESS;
+    }
+
+    /**
+     * Commit the transaction.
+     * The commit call begins a Corfu transaction at the specified snapshot or at the latest snapshot
+     * if no snapshot is specified, applies all the updates and then ends the Corfu transaction.
+     * The commit returns successfully if the transaction was committed.
+     * Otherwise this throws a TransactionAbortedException.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
+        long commitAddress;
+        try { // batch up all the operations and apply them in a new transaction started here.
+            txBeginMaybe();
+            operations.forEach(Runnable::run);
+        } finally {
+            commitAddress = endTxn();
+        }
+        return commitAddress;
+    }
```

Comment:
Can we just do 

```java
public long commit() {
        long commitAddress;
        txBeginMaybe();
        operations.forEach(Runnable::run);
        operations.clear();
        return this.objectsView.TXEnd();
}

Because anyway user needs to start transaction in try finally block to prevent there code throwing exception and not cleaning up resource.
```

Reply:
Just wanted the try {} finally{} block for collecting metrics on how many transactions are aborting. If we just call TXEnd() without any try catch block then we can't capture this metric right?

Reply:
Then the try should be on endTxn() right? But here the try is on applying the inmemory operations. Since user code can throw exceptions anytime even before commit, which is effectively an abort. So I guess we just need to collect the metrics on aborts thrown from underlying runtime.getObjectsView.TXEnd().

Reply:
but will add it in the metrics PR not here.

Reply:
@hisundar I think the current usage of try finally block here might have some problem. If exception happens in operations.forEach(Runnable::run), then endTxn() will still be called. However endTxn() is not just resource cleanup, it also includes doing actual commit, which means you could exceute part of the operations list and hit an exception and then trying to commit that partial transaction, right?